jwst/extract_1d/extract.py

```diff
@@ -2562,14 +2563,25 @@ def run_extract1d(input_model, refname, smoothing_length, bkg_order,
         obtained by iterating over a SourceModelContainer.  The default
         is False.
 
+    apcorr_ref_name : str
+        Name of the APCORR reference file. Default is None
+
     Returns
     -------
     output_model : data model
         A new MultiSpecModel containing the extracted spectra.
-    """
 
+    """
     # Read and interpret the reference file.
-    ref_dict = load_ref_file(refname)
+    ref_dict = load_ref_file(extract_ref_name)
+
+    apcorr_table = apcorr_sizeunits = None
+    if apcorr_ref_name is not None or apcorr_ref_name != 'N/A':
+        with fits.open(apcorr_ref_name) as ap_ref:
```

Comment:
Don't we have data models defined for apcorr ref files and hence should/could be using datamodels.open() here, rather than direct FITS I/O? We try to avoid any direct I/O as much as possible and let datamodels handle it.

Reply:
I assume this is because datamodels table reader drops units defined in TUNIT on the floor?  This really should be fixed in the datamodels infrastructure.

The bug is somewhere in `datamodels.util.gentle_asarray()`:

https://github.com/spacetelescope/jwst/blob/master/jwst/datamodels/util.py#L326-L377

The `unit` info is the the column, but we just use the `dtype`.  There's no unit info in the `dtype`.  The above code should probably be refactored to use `column` objects instead of `dtype`.

Reply:
There's even an open issue for this

https://github.com/spacetelescope/jwst/issues/3979

Tackle in an upcoming sprint?  🤞

Reply:
We added a dedicated keyword to the headers of the apcorr ref files to store the units, separate from TUNITn, in order to work around this problem. So as long as there's a meta data entry for that special keyword in the datamodel schema for the ref file, we should be good (i.e. the data and units will both be available in the model).

Reply:
> There's even an open issue for this
> 
> #3979
> 
> Tackle in an upcoming sprint? 🤞

Are you volunteering?

Reply:
Yep, so originally I was using `jwst.datamodels.open`, but we were having issues with consistently being able to extract the units for one of the columns. Additionally, some of the reference files don't contain a `DATAMODL` keyword, and I wasn't sure how to get around that without writing a function with a hardcoding mapping of some kind. Thoughts?

Reply:
You can work around the missing `DATAMODL` keyword by explicitly opening with the required/desired data model, e.g. `jwst.datamodels.NrcWfssApcorrModel`, but that of course has the added burden of having separate open calls for every mode of data. We should just enforce the presence of `DATAMODL` in the reference files.

Reply:
@hbushouse that's what I was thinking as well and I think it would allow for trimming number of functions that do exactly that for other steps.